src/red_black_tree.rs

```diff
@@ -933,6 +959,93 @@ impl<
     }
 }
 
+pub struct RedBlackTreeDrainFilter<
+    'a,
+    K: Debug + PartialOrd + Ord + Copy + Clone + Default + Pod + Zeroable,
+    V: Default + Copy + Clone + Pod + Zeroable,
+    P: for<'p> Fn(&'p K, &'p V) -> bool,
+    const MAX_SIZE: usize,
+> {
+    tree: &'a mut RedBlackTree<K, V, MAX_SIZE>,
+    fwd_stack: Vec<u32>,
+    fwd_ptr: u32,
+    fwd_node: Option<u32>,
+    _rev_stack: Vec<u32>,
+    _rev_ptr: u32,
+    rev_node: Option<u32>,
+    terminated: bool,
+    /// Keeps addr's of nodes with predicate = true, to remove upon dropping.
+    /// It is possible to instead drop during iteration and update fwd_ptr & fwd_stack
+    remove: Vec<u32>,
+    predicate: P,
+}
+
+impl<
+        'a,
+        K: Debug + PartialOrd + Ord + Copy + Clone + Default + Pod + Zeroable,
+        V: Default + Copy + Clone + Pod + Zeroable,
+        P: for<'p> Fn(&'p K, &'p V) -> bool,
+        const MAX_SIZE: usize,
+    > Drop for RedBlackTreeDrainFilter<'a, K, V, P, MAX_SIZE>
+{
+    fn drop(&mut self) {
```

Comment:
The gist of the speed up here is that the traversal remove nodes is O(K) instead of O(K log N) right?

Reply:
Actually still O(K log N) but you save a log N on the find

Reply:
Essentially, with `remove` you have to traverse the tree from the root to the entry to re-find the node addr of the entry you just found. 

If you iterate through the elements and record the addr instead of the keys, you don't have to do that traversal to remove the entry. That's the gist